### PR TITLE
Uses reader Topics to populate the new Reader Tab Bar

### PR DESCRIPTION
**Pull request by Gio2018**
Ref #13580

To test:

- build/run and login with an account that has one or more teams, and possibly some lists
- Go to the Reader
- Verify that the first four items in the tab bar are the default items, in this order
  - Following
  - Discover
  - Likes
  - Saved
- Verify that teams (and lists, if you have any) are displayed after the four default items and are listed in alphabetical order

#### Notes: 
1. in this implementation, the viewModel talks directly to CoreData and ReaderTopicService. I'm keen to abstract this part away, in a different object, but I also don't want to over-engineer the code. Reviewers please share your opinion.
2. At this stage, it is expected that some UI tests may fail. Feature flag will be disabled before merging into develop.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by Gio2018** [comment]
> I'm also seeing some visual issues during the initialization of the view, where the filter bar initially appears at the bottom of the view and then pops to the top. Have you noticed this?

Thanks for pointing this out! I believe it was due to the height constraint of `FilterTabBar` not being reactivated after changing the height value. It should be fixed now.

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabView.swift [hunk 1]
```diff
@@ -72,6 +69,19 @@ extension ReaderTabView {
         tabBar.tabBarHeight = Appearance.barHeight
         WPStyleGuide.configureFilterTabBar(tabBar)
         tabBar.addTarget(self, action: #selector(selectedTabDidChange(_:)), for: .valueChanged)
+
+        viewModel.fetchReaderMenu() { [weak self] sections in
+            guard let sections = sections, let self = self else {
+                return
+            }
+            UIView.animate(withDuration: Appearance.tabBarAnimationsDuration) {
```
[the remark this answers is not in the document]
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabView.swift:76]: we could, but I would still use the inner scope of the `fetchReaderMenu()` closure (instead of having the convenience method to point to the outer scope). To do so, either I pass three parameters (self.tabBar, self.buttonsStackView and sections) or the inner reference to `self` (aka `ReaderTabView`) along with the sections, and both seem a little overkill for a few lines of code.

File: WordPress/Classes/ViewRelated/Reader/ReaderHelpers.swift [hunk 2]
```diff
@@ -233,3 +233,43 @@ import WordPressShared
...
+            return NSLocalizedString("Following", comment: "Title of the Following Reader tab")
+        } else if topicIsLiked(topic) {
+            return NSLocalizedString("Likes", comment: "Title of the Likes Reader tab")
+        }
+        return topic.title
+    }
+    /// Sorts the default tabs according to the order [Following, Discover, Likes], and adds the Saved tab
+    class func rearrange(items: [ReaderTabItem]) -> [ReaderTabItem] {
+        var mutableItems = items
+        mutableItems.sort {
```
[the remark this answers is not in the document]
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Reader/ReaderHelpers.swift:251]: I agree that is not super readable; I also agree that 1 could be tricky. I have added some comments in that method, hoping that they add some clarity to it.

[the remark this answers is not in the document]
Comment (bjtitus) [reply WordPress/Classes/ViewRelated/Reader/ReaderHelpers.swift:251]: I think this may be more refactoring than necessary for this PR, but here are some thoughts I had looking through this:

It seems like we could replace these individual methods with an enum for each topic type:

```swift
enum ReaderTopicType {
    case following
    case discover
    case liked
    
    var path: String {
        switch self {
        case .following:
            return "/read/following"
        case .discover:
            return "/read/sites/53424024/posts"
        case .liked:
            return "/read/liked"
        }
    }
}
```

The sort code could then take a list of topics for sorting and be more general. Although, this might be harder to reason about in the end.

```swift
let sortedTopics: [ReaderTopicType] = [.following, .discover, .liked]

var mutableItems = items

mutableItems.sort { (leftTopic, rightTopic) in
    let lhsIndex = sortedTopics.firstIndex { type in
        return leftTopic.topic?.path.hasSuffix(type.path) == true
    }
    let rhsIndex = sortedTopics.firstIndex { type in
        return rightTopic.topic?.path.hasSuffix(type.path) == true
    }
    
    if let lhsIndex = lhsIndex, let rhsIndex = rhsIndex {
        /// Sort from `sortedTopics`
        return lhsIndex < rhsIndex
    } else if lhsIndex != nil || rhsIndex != nil {
        /// Sort any topics from`sortedTopics` over those not in the list
        return lhsIndex != nil
    } else if leftTopic.topic?.type == rightTopic.topic?.type {
        /// Sort any others topics alphabetically, grouped by topic type
        return leftTopic.title < rightTopic.title
    } else {
        return true
    }
}

// fourth item: Saved. It's manually inserted after the sorting
mutableItems.insert(ReaderTabItem(title: NSLocalizedString("Saved", comment: "Title of the Saved Reader Tab")), at: 3)
return mutableItems
```

[the remark this answers is not in the document]
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Reader/ReaderHelpers.swift:251]: This is a valid point, and I also agree that this kind of refactoring could go in a separate PR. Another thought that I had is that, if the default topics are not going to change for the foreseeable future, maybe we don't even need to fetch them? If we all agree, I'd address this task on a separate PR, once the main navigation is in place.

**Comment by bjtitus** [comment]
I think I'm still seeing the issue that @frosty posted with the odd animation of the filter bar:

FilterBarAnimationPopIn

**Comment by Gio2018** [comment]
> I think I'm still seeing the issue that @frosty posted with the odd animation of the filter bar:
> 

I believe the issue pointed out by frosty was different:

> I'm also seeing some visual issues during the initialization of the view, where the filter bar initially > appears at the bottom of the view and then pops to the top

What your gif is showing, is the animation of the tab bar being populated, which we can also remove, if we feel we have to, but it's a different thing

**Comment by Gio2018** [comment]
> This looks good, although I'm also seeing the initial layout animation as Brandon. I think we should probably remove the animation block from `populateTabBar`.

Done! I also took the occasion to test this on simulated bad network conditions (3G and Edge) and it looks reasonably fast.

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabView.swift [hunk 3]
```diff
@@ -78,13 +78,11 @@ extension ReaderTabView {
     }
 
     private func populateTabBar(with items: [ReaderTabItem]) {
-        UIView.animate(withDuration: Appearance.tabBarAnimationsDuration) {
-            self.tabBar.items = items
-            guard let tabItem = self.tabBar.items[self.tabBar.selectedIndex] as? ReaderTabItem else {
-                return
-            }
-            self.buttonsStackView.isHidden = tabItem.shouldHideButtonsView
+        self.tabBar.items = items
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabView.swift:81]: One last nitpick – we could remove the `self` references from here now it's no longer in a block :)

**Comment by Gio2018** [comment]
> I think this looks good 👍
> 
> A little more could probably be done to improve the interaction with safe areas. Here are the two changes I used:
> 
> Added to `setupButtonsView`:
> 
> ```swift
> buttonsStackView.isLayoutMarginsRelativeArrangement = true
> ```
> 
> Changed in `activateConstraints`:
> 
> ```swift
> pinSubviewToAllEdges(mainStackView)
> ```
> 

Yup, this makes total sense, thanks for pointing this out! I'll include these two changes too
